Fixes #11458 - Override inline for class parameters

tbrisker opened the pull request:
This implements override inline for smart class parameters and smart
variables in host and hostgroup edit forms.
Global variable override inline will be done in a seperate PR.

Authors:
    Ori Rabin orrabin@redhat.com
    Tomer Brisker tbrisker@redhat.com

tbrisker committed:
Fixes #11458 - Override inline for class parameters

This implements override inline for smart class parameters and smart
variables in host and hostgroup edit forms.
Global variable override inline will be done in a seperate PR.

Authors:
	Ori Rabin <orrabin@redhat.com>
	Tomer Brisker <tbrisker@redhat.com>

domcleal commented on app/helpers/lookup_keys_helper.rb:
```diff
@@ -74,35 +74,39 @@ def overridable_lookup_keys(klass, host)
       end
     else
-      diagnostic_helper = popover('', _("<b>Description:</b> %{desc}<br><b>Type:</b> %{type}<br> <b>Matcher:</b> %{matcher}") % { :desc => key.description, :type => key.key_type, :matcher => matcher}, :data => { :placement => 'top' })
+      diagnostic_helper = popover('', _("<b>Description:</b> %{desc}<br/><b>Type:</b> %{type}<br/> <b>Matcher:</b> %{matcher} <br/><b>Value:</b> %{original_value}") % { :desc => key.description, :type => key.key_type, :matcher => matcher, :original_value => original_value }, :data => { :placement => 'top' }, :title => _("Original value info"))
```
Is it worth calling this the inherited or original value, rather than value?

tbrisker replied:
Note that the popover title states this is all info related to the original value, do you think this needs more clarificaion?

domcleal commented:
1) When editing a host, global parameters inherited from a host group etc. no longer are readonly - they have editable text boxes.

2) Moving the use_puppet_default help text to the top instead of next to every checkbox is good, but consider adding a tooltip to the checkbox just so you know what it is.  It doesn't need the full explanation, just the title.

3) The value fields are a lot more narrow than the previous overridden value fields, and slightly narrower than the inherited value fields used to be.  I think this will be OK because of the full screen button, but can you see if we can reclaim any more horizontal space?

Perhaps from the Use Puppet default column or even going back to icons instead of override/remove buttons?

4) It would be good to highlight overridden fields more, as this was a benefit for some users of the old layout.  The background colour of the text field and presence of the checkbox are the main visual clues right now, but I'd suggest also changing the border colour of the value text boxes (perhaps to blue?) or making the parameter name bold.

Else probably something for a future PR, some links/buttons at the top to filter/only view overridden parameters.

5) Can the remove/override buttons be the same width?

6) I don't think the PR makes it any worse than it currently is, but FYI I'm bumping into this nasty issue when overriding YAML lookup keys inherited from host groups: http://projects.theforeman.org/issues/12284

Looks really good, I'm impressed!

tbrisker commented:
@domcleal I have updated to original commit to address your first comment, and added the corrections for the rest of the comments to a separate commit which can be squashed later if accepted, just to allow simple reversion if we decide to leave it as is and address those issues in the future.

dLobatog commented:
@tbrisker I've taken a look - not properly reviewed it at all, but from the UI/UX point of view.
- Every time I select an environment, a new header Puppet class parameters is added. See it in action 

![demo gif parameters duplicate environment](http://i.imgur.com/NuCiicO.gif)
- The yellow background color for overrides is doesn't fit very well I would say. Also for parameters without value, it shows the warning in yellow by default (when it's not overridden), and the border in red?

![yellow override](http://i.imgur.com/bOLNa4S.png). 

What @domcleal suggested seems to look better, I would also suggest to display a small popover 'Overridden' when you hover the mouse on that table row:

![screenshot from 2015-11-05 13-39-50](https://cloud.githubusercontent.com/assets/598891/10968610/bd5c6ad0-83c2-11e5-82bd-147cdec13888.png)
- The override button looks very out of place, and it shows up on the Name cell. Why not adding it to the value textarea? Or even the actions column is better than this one I'd say.
  
  ![override button](http://i.imgur.com/nN2BtYd.png)
- Long Puppet class names are truncated but it'd be nice to see the whole name on mouse hover.
- 'Use puppet default' doesn't work at all? It doesn't show checkboxes.

Let me know what you think :ok_hand:

tbrisker commented:
@dLobatog I addressed most of your comments. 
For the values that have an empty default - if they are not required they show a yellow warning and the border is yellow, if they are required the border is red. I have changed it now so that clicking on override will remove the border.
Long class names should show whole name on hover, however the tooltip is attached to the `<td>` so it might appear higher if the class has multiple parameters. 
'Use puppet default' checkbox is only displayed for smart class parameters, not smart variables as they have no puppet default.

dLobatog commented:
@tbrisker Thanks! A few more:
- I don't see any tooltips on hover
- Top left and bottom left border-radius of the `override` button should be 0px to blend in with the full screen button.
- Is this the way required empty values are supposed to look? It looks a bit strange in my opinion, the warning sign is enough in my opinion. I can't see any guidelines for highlighting elements in a table on [Patternfly](patternfly.org/) @kybaker any hints?
  ![](http://i.imgur.com/qm10ymn.png) 

This doesn't have a lot to do with the PR, but what do you think about softening the borders of the input-group-addon and the textarea? I might do it on a PR after this unless you want to include it here, it's fairly easy.

![screenshot from 2015-11-06 19-20-21](https://cloud.githubusercontent.com/assets/598891/11005053/efc6baee-84bb-11e5-90ca-5d2d6b0827cc.png)

kybaker commented:
@tbrisker In PatternFly if a user does not fill a required field or a user enters invalid data the border becomes red [1]. If the goal is to just draw the users attention and to differentiate the field visually I think the treatment above might be okay.

As far as softening the borders - The only time softening the borders is appropriate is when a field is disabled. Even when the field is disabled it is important to have enough contrast so the field does not just look like a rectangle. [1]

[1] https://www.patternfly.org/widgets/#forms

tbrisker commented:
@dLobatog I have removed the color highlighting and fixed the buttons, improved the code a bit from previous commits and squashed it all together.
I am not sure why you aren't getting tooltips, they seem to work fine for me - could you please give more info on this matter?
As for the softer borders I tend to agree with @kybaker that it doesn't look very good in this situation.

domcleal commented:
> I don't see any tooltips on hover

I think that's existing bug that might happen if you just added the Puppet class.  If you save the host and come back, they ought to work.

dLobatog commented:
Reviewed 6 of 20 files at r1.
Review status: 6 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[app/helpers/application_helper.rb, line 64 \[r1\]](https://reviewable.io:443/reviews/theforeman/foreman/2858#-K2kn0_SkbtJs5BPNqNh)** ([raw file](https://github.com/theforeman/foreman/blob/30b48cc2fe0c5f0b3c6fb7bcac595e89cb266802/app/helpers/application_helper.rb#L64)):</sup>
I s this change (& the equivalent on javascript) related to the issue? 
<!-- Reviewable comment -K2kn0_SkbtJs5BPNqNi:1838547876 -->

---

<sup>**[app/models/concerns/host_common.rb, line 45 \[r1\]](https://reviewable.io:443/reviews/theforeman/foreman/2858#-K2koNRKjzW1qb-mPUJV)** ([raw file](https://github.com/theforeman/foreman/blob/30b48cc2fe0c5f0b3c6fb7bcac595e89cb266802/app/models/concerns/host_common.rb#L45)):</sup>
Under what circumstances would the other `if` fail? 
<!-- Reviewable comment -K2koNRKjzW1qb-mPUJW:-671248777 -->

---

<sup>**[app/views/provisioning_templates/_combinations.html.erb, line 5 \[r1\]](https://reviewable.io:443/reviews/theforeman/foreman/2858#-K2kokb9YeZWxibKg40P)** ([raw file](https://github.com/theforeman/foreman/blob/30b48cc2fe0c5f0b3c6fb7bcac595e89cb266802/app/views/provisioning_templates/_combinations.html.erb#L5)):</sup>
I'm a bit confused, sorry, why this change?
<!-- Reviewable comment -K2kokbAmOn33jGOVNXZ:774834265 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/theforeman/foreman/2858)
<!-- Sent from Reviewable.io -->

dLobatog commented on app/helpers/application_helper.rb:
```diff
@@ -61,7 +61,7 @@ def link_to_add_fields(name, f, association, partial = nil, options = {})
       render((partial.nil? ? association.to_s.singularize + "_fields" : partial), :f => builder)
     end
-    link_to_function(name, ("add_fields(this, \"#{association}\", \"#{escape_javascript(fields)}\")").html_safe, add_html_classes(options, "btn btn-success") )
+    link_to_function(name, ("add_fields('#{options[:target]}', '#{association}', '#{escape_javascript(fields)}')").html_safe, add_html_classes(options, "btn btn-success") )
```
Is this change (& the equivalent on javascript) related to the issue?

tbrisker replied:
This is so that the element to add to can be specified to something other then `this` - I needed this change to make the global parameter to align nicely with the puppet class parameters. (https://github.com/theforeman/foreman/pull/2858/files#diff-b74a9e3cff7e9e6af6614aea6e50e3a5R16)

dLobatog commented on app/views/provisioning_templates/_combinations.html.erb:
```diff
@@ -2,5 +2,5 @@
   <% end %>
-  <p><%= link_to_add_fields('+ ' + _("Add combination"), f, :template_combinations, "provisioning_templates/combination") %></p>
 <% end %>
+<p><%= link_to_add_fields('+ ' + _("Add combination"), f, :template_combinations, "provisioning_templates/combination", :target => '#template_combination') %></p>
```
I'm a bit confused, sorry, why this change?

tbrisker replied:
This is to conform to the change made above, in ApplicationHelper#link_to_add_fields

dLobatog commented on app/models/concerns/host_common.rb:
```diff
@@ -41,7 +41,8 @@ module HostCommon
         attr = attribute.dup
-        if id = attr.delete(:id)
+        id = attr.delete(:id)
+        if id.present?
```
Under what circumstances would the other `if` fail?

tbrisker replied:
when id=''

dLobatog commented on app/helpers/lookup_keys_helper.rb:
```diff
@@ -72,35 +72,41 @@ def overridable_lookup_keys(klass, host)
-    value, matcher = value_for_key ? [value_for_key[:value], "#{value_for_key[:element]} (#{value_for_key[:element_name]})"] : [key.default_value, _("Default value")]
     original_value = key.value_before_type_cast value
-    no_value = value.nil? && key.lookup_values.find_by_match("fqdn=#{host.fqdn}")
+    overriden_value = lookup_value.try(:value)
```
When would `lookup_value` not have a `:value` method?

tbrisker replied:
`lookup_value` is `nil` when the key is not overridden for this fqdn/hostgroup.

dLobatog commented on app/helpers/lookup_keys_helper.rb:
```diff
@@ -72,35 +72,41 @@ def overridable_lookup_keys(klass, host)
     original_value = key.value_before_type_cast value
-    no_value = value.nil? && key.lookup_values.find_by_match("fqdn=#{host.fqdn}")
+    overriden_value = lookup_value.try(:value)
+    no_value = value.to_s.empty? && overriden_value.to_s.empty?
```
`value.blank? && overriden_value.blank?` should be enough, see http://apidock.com/rails/Object/blank%3F

tbrisker replied:
Just realized why we did this - the value might be `false`, in which case `blank?` will return true and `to_s.empty?` will return false. reverting this back to the way it was.

dLobatog commented on app/helpers/lookup_keys_helper.rb:
```diff
@@ -72,35 +72,41 @@ def overridable_lookup_keys(klass, host)
-    diagnostic_helper = popover('', _("<b>Description:</b> %{desc}<br><b>Type:</b> %{type}<br> <b>Matcher:</b> %{matcher}") % { :desc => key.description, :type => key.key_type, :matcher => origin}, :data => { :placement => 'top' })
-    parameter_value_content("value_#{key.key}", original_value, :popover => diagnostic_helper)
-  end
+  def key_with_diagnostic(obj, key, index, lookup_value, value_hash = nil)
```
I find this considerably harder to read than what we had before, one method for hostgroup and another for hosts. What do we gain out of merging them if we have to constantly check what object are we using inside?

tbrisker replied:
This is to reduce code duplication between the two different versions, most of the code is the same for both.

dLobatog commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,48 @@
+              <%= link_to_function(icon_text('edit'), "override_class_param(this)", :title => _("Override this value"),
+                                   :'data-tag' => 'override', :class =>"btn btn-default btn-md #{'hide' if overridden}")%>
+              <%= link_to_function(icon_text('remove'), "override_class_param(this)", :title => _("Remove this override"),
+                                   :'data-tag' => 'remove', :class =>"btn btn-default btn-md #{'hide' if !overridden}")%>
```
`if !` -> `unless`. 
In any case, how about moving this to a helper `override_button` - including the check for `can_edit`?

dLobatog commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,48 @@
+  <% overridden = key.overridden?(obj) %>
+  <% disabled = can_edit && !overridden %>
+  <% lookup_value = obj.lookup_values.detect{|v| v.lookup_key_id == key.id} %>
+  <% puppet_default = lookup_value.try(:use_puppet_default) %>
```
This is only used once below, can you call it directly? I think you won't even need to call `.try` since you have a check `if key.type == "PuppetclassLookupKey"` there.

tbrisker replied:
Moved, but still need try because `lookup value` may be `nil`.

dLobatog commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,48 @@
+  <% disabled = can_edit && !overridden %>
+  <% lookup_value = obj.lookup_values.detect{|v| v.lookup_key_id == key.id} %>
+  <% puppet_default = lookup_value.try(:use_puppet_default) %>
+  <% error = lookup_value.try(:errors) %>
```
Afaik, no need to call `.try`? `:errors` is always there on ActiveRecord models

tbrisker replied:
ditto

dLobatog commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,48 @@
+          </span>
+        </div>
+        <% if error.present? %>
+          <%= content_tag(:span, error.messages[:value].try(:to_sentence), :class => "help-block") %>
```
Why `.try`? When is messages[:value] empty or something similar?

tbrisker replied:
There is a case when cloning a hostgroup that throws an error on parameter overrides because the match is invalid before saving (and on the way, highlighting all the overridden values).
In that case, `error.messages == {:match=>["is invalid"]}`, and `error.messages[:value]` is nil.

dLobatog commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,48 @@
+      </td>
+      <td class="ca">
+        <%= check_box(value_prefix, :use_puppet_default,
+                      :value    => lookup_value.try(:id),
```
Ditto, `:id` is always available even if it's nil

tbrisker replied:
ditto

dLobatog commented:
@tbrisker @orrabin Thanks, works pretty well now. I have a few questions about the code, I'm not sure I fully understand `key_with_diagnostic` and `_class_parameters`. Please see the notes inline.

dLobatog commented:
I tried to fix a bit issues with `key_with_diagnostic` and `_class_parameters` on https://github.com/tbrisker/foreman/pull/2 , I'll comment in line there

dLobatog commented on app/helpers/lookup_keys_helper.rb:
```diff
@@ -68,39 +68,122 @@ def validator_type_selector(f)
+      value_for_key = value.try(:[], lookup_key.key)
+      if value_for_key.present?
+        [value_for_key[:value],
+          "#{value_for_key[:element]} (#{value_for_key[:element_name]})"]
```
Indentation's off 1 space here, rubocop's unhappy

tbrisker replied:
fixed

dLobatog commented on app/helpers/lookup_keys_helper.rb:
```diff
@@ -68,39 +68,122 @@ def validator_type_selector(f)
       else
-        diagnostic_class = 'warning'
-        diagnostic_helper = popover('', _("Optional parameter without value.<br/><i>Won\'t be given to Puppet.</i> <br><br><b>Description:</b> %s") % key.description, :icon => "warning-sign")
+          [lookup_key.default_value, _("Default value")]
```
Same here, indentation is off 2 spaces

tbrisker replied:
fixed

dLobatog commented on app/models/lookup_keys/lookup_key.rb:
```diff
@@ -144,6 +144,12 @@ def overridden?(host)
     host.lookup_values.any? { |lv| lv.lookup_key_id == id }
   end
 
+  def overridden_value(host)
```
If you are going to add this change, 3 lines above you can have the previous method work in terms of this one:

``` ruby
overridden_value(host).present?
```

tbrisker replied:
Changed. Also changed `host` to `obj` since it can be also a host group and in the future possible other objects, and added comment on why we do it by iteration instead of by DB query.

dLobatog commented on app/helpers/lookup_keys_helper.rb:
```diff
@@ -68,39 +68,122 @@ def validator_type_selector(f)
+  end
+
+  def value_matcher(obj, lookup_key)
+    if obj.class.model_name == "Host"
```
You have a method `parameters_receiver` below to check the parameters_receiver based on params rather than checking the class, both are code smells though, this needs a very deep refactor (not for this PR, but we should keep it in mind). Maybe you can move this logic to the model as I try to explain below.

tbrisker replied:
There was a mistake in the `parameters_receiver` function that when `params[:host] == {}` `.present?` would return false and cause a test to fail. Changed it to `params.has_key?` instead and changed this as well.

dLobatog commented on app/helpers/lookup_keys_helper.rb:
```diff
@@ -68,39 +68,122 @@ def validator_type_selector(f)
+          [lookup_key.default_value, _("Default value")]
       end
+    else #hostgroup
+      obj.inherited_lookup_value(lookup_key)
```
How about moving these to their own helpers, or the model even? 
Ideally we should have a HostForm and HostgroupForm, or  call `value_matcher(obj, lookup_key)` and it should respond from the proper helper. The only reason why that doesn't work is because we load _all_ helpers on _all_ views and there's a name conflict. 

But moving this logic to the model itself would be fine and respond well. You would just call `obj.value_matcher(lookup_key)` where needed and it responds properly depending on what `obj` is, without type checking.

tbrisker replied:
The reason this is done in the helper is that `value_hash_cache` is only created in the helper when it is accessed. I chose not to split this into the different helpers as you suggested as it caused the code to be more difficult to follow and understand.

dLobatog commented:
@tbrisker @orrabin Aside from the class checking on `value_matcher` I see this buggy behavior:
- It clears the original value if I decide to override it and then I change my mind ![](http://i.imgur.com/5XuSW79.gif)

Sorry for holding this one up, but I think it's important we keep this part of the code as clear as possible since 'it needs refactoring' but it may take a long time before someone actually goes ahead and does it. Right now aside from the 2 issues above I say :+1: to the rest of the PR, thanks !

tbrisker commented:
@dLobatog I addressed all your comments and fixed the js bug that cleared the field.

dLobatog commented:
Merged as 1e1339f5683adb133234983c28853943d573b312, thanks @tbrisker & @orrabin , nice work!
